packages/rich-text/src/to-tree.js

```diff
@@ -0,0 +1,93 @@
+/**
+ * Internal dependencies
+ */
+
+import { split } from './split';
+
+export function toTree( value, multilineTag, settings ) {
+	if ( multilineTag ) {
+		const { createEmpty, append } = settings;
+		const tree = createEmpty();
+
+		split( value, '\u2028' ).forEach( ( piece, index ) => {
+			append( tree, toTree( piece, null, {
+				...settings,
+				tag: multilineTag,
+				multilineIndex: index,
+			} ) );
+		} );
+
+		return tree;
+	}
+
+	const {
+		tag,
+		multilineIndex,
+		createEmpty,
+		append,
+		getLastChild,
+		getParent,
+		isText,
+		getText,
+		remove,
+		appendText,
+		onStartIndex,
+		onEndIndex,
+	} = settings;
+	const { formats, text, start, end } = value;
+	const formatsLength = formats.length + 1;
+	const tree = createEmpty( tag );
+
+	append( tree, '' );
+
+	for ( let i = 0; i < formatsLength; i++ ) {
```

Comment:
Since `formats` is a sparse array, would it be worth considering using `forEach` to skip over the unformatted segments?

>forEach calls callbackfn _[...]_ only for elements of the array which actually exist; it is not called for missing elements of the array.

https://www.ecma-international.org/ecma-262/5.1/#sec-15.4.4.18

When a format is encountered and at the end of the text, we could simply append the unformatted text in a single go. This would also be much more performant since `append` is a non-trivial operation.

Benchmarks:

- [`forEach` is 84x faster iterating an array of 100 empty values (e.g. a paragraph with 100 characters and no formatting)](https://jsbin.com/yakahig/2/edit?html,js,console,output)
- [`append` is 17x faster when appending in one go, rather than one character at a time](https://jsbin.com/recoven/edit?html,js,console,output)

Reply:
Yes, maybe we can do this. I'd have to explore it, so far didn't have the time. Formats that stay equal could also be appended in one go. Also note that `appendData`, not `append`, which probably doesn't make it any faster.